# PYTHON MIDTERM EXAM
# Bulahan, Joren Jay L.
Write roll() -> int to simulate a fair die. Then simulate 100,000 rolls of two dice and estimate the probability that their sum equals 7. Compare the empirical result to the exact probability. Comment on random seeding and reproducibility.
Anchor concepts: random module, simulation, averages
Reference: Lecture on Monte Carlo methods, reproducibility, and probability estimation

# PROMPT (paste verbatim)
Write a Python function roll() that simulates a fair die (1–6). Then simulate 100,000 rolls of two dice and estimate the probability that their sum equals 7. Compare to the exact probability. Mention random seeding and reproducibility.


In [5]:
import random

def roll():
    return random.randint(1, 6)

def simulate():
    count = 0
    trials = 100000
    for _ in range(trials):
        if roll() + roll() == 7:
            count += 1
    probability = count / trials
    print(f"Estimated probability of sum=7: {probability:.4f}")

simulate()

Estimated probability of sum=7: 0.1652


# Correctness
- roll() uses random.randint(1, 6) — simulates a fair die.
-  Simulates 100,000 trials and counts how often the sum is 7.
-  Computes empirical probability correctly.
# Time & Space Complexity
- Time: O(n) — each trial involves two rolls and a comparison.
- Space: O(1) — only a counter and fixed variables.
# Robustness
-  No random seed — results vary each run.
-  No docstrings or type hints.
-  Handles large number of trials efficiently.
# Readability / Style
-  Clear function names and structure.
-  No comments or explanation of expected probability.
-  No modularity — could separate simulation and reporting.
# Faithfulness to Lecture
-  Uses random.randint() for uniform distribution.
-  Monte Carlo simulation matches lecture examples.
-  Doesn’t mention reproducibility via random.seed().

# Improved Code

In [7]:
import random

def roll() -> int:
    """Simulate a fair six-sided die roll."""
    return random.randint(1, 6)

def simulate_dice_sum(target_sum: int = 7, trials: int = 100_000, seed: int = 42) -> float:
    """
    Simulate rolling two dice `trials` times and estimate the probability
    that their sum equals `target_sum`. Uses a fixed seed for reproducibility.
    """
    random.seed(seed)
    count = sum(1 for _ in range(trials) if roll() + roll() == target_sum)
    return count / trials

# ✅ Run simulation
if __name__ == "__main__":
    estimated = simulate_dice_sum()
    exact = 6 / 36  # There are 6 combinations that sum to 7 out of 36 total
    print(f"Estimated probability of sum=7: {estimated:.4f}")
    print(f"Exact probability: {exact:.4f}")

Estimated probability of sum=7: 0.1646
Exact probability: 0.1667


# Discussion: Probability of Sum = 7
- Total outcomes for two dice: 6 \times 6 = 36
- Favorable outcomes for sum = 7: (1,6), (2,5), (3,4), (4,3), (5,2), (6,1) → 6 outcomes
- Exact probability: 6 / 36 = 0.1667
# Random Seeding
- random.seed(42) ensures reproducibility — same results every run
- Useful for debugging, testing, and consistent grading

# Lecture Concepts Cited
- Monte Carlo simulation: Empirical estimation via repeated trials
- Random module: randint() for uniform discrete sampling
- Reproducibility: random.seed() fixes randomness
- Probability theory: Comparing empirical vs. exact outcomes
